ui: add ability to cancel statement diagnostics to the frontend

THardy98 opened the pull request:
**Note**: This PR was split into two PRs: #75733, #75732.

Previously, there did not exist an option to cancel a running request
for statement diagnostics, this change provides a cancellation option.

The statements page now has an ellipsis dropdown button where the
download bundle links used to appear. The dropdown contains a
cancellation button (if there is a currently waiting diagnostic request)
and the download links to previously completed diagnostic requests.

The statement details page now has a "Cancel request" button while a
diagnostic request is not completed (i.e. in the "Waiting" status).

The statement diagnostics history page on db-console also has a "Cancel
Request" button while a diagnostic request is not completed (i.e. in the
"Waiting" status).

Added a small code change to the cancel statement diagnostics request
API on the backend. Previously, the frontend passed a statement
fingerprint to the API to update the corresponding statement diagnostic
request to an expired state. Now we pass the statement diagnostic ID
directly. This resolves a SQL parsing error that occurs when a statement
fingerprint contains single-quotes (i.e. when a statement fingerprint
contains placeholders).

Added a small code change in the statement diagnostic "Activate" modal.
The expiry time picker allowed for values that were <1, the code change
fixes this bug.

Note: Follow up [issue](https://github.com/cockroachdb/cockroach/issues/75559) regarding alerts & notifications in db-console and managed-service.

**DB Console Changes**:
https://www.loom.com/share/f76694148f6f4fcdb129665eb519eb4b

**CC Console Changes - this is using intrusion & console with fake backends**
https://www.loom.com/share/06bdc2455b5c49d6ba0d23cc53c7b460

Release note (ui change): Added an option to cancel a running request
for statement diagnostics.

matthewtodd reviewed (changes requested):
Nice work, Thomas! Just a couple of questions below.

> Reviewed 32 of 32 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-:-Mtifa7e-Vy8ilwxVqNO:b-2u266y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @THardy98)

___
*[pkg/server/serverpb/status.proto, line 1282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtidRyl7FDAsA-zgZB_:-MtidRyl7FDAsA-zgZBa:bkub25l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/server/serverpb/status.proto#L1282)):*
> ```protobuf
> 
> message CancelStatementDiagnosticsReportRequest {
>   int64 request_id = 1;
> ```

In general, it's not safe to switch out fields like this. But I may be missing something from the context: is this new code, i.e., are there no clients yet?
___
*[pkg/sql/stmtdiagnostics/statement_diagnostics.go, line 405 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtiduw4Bz6klaIim6BJ:-Mtiduw4Bz6klaIim6BK:b-pm5n21) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/sql/stmtdiagnostics/statement_diagnostics.go#L405)):*
> ```Go
> 		// allow any queries that are currently being traced for this request to
> 		// write their collected bundles.
> 		fmt.Sprintf("UPDATE system.statement_diagnostics_requests SET expires_at = '1970-01-01' "+
> ```

While we're here, it's safer (from a SQL-injection perspective) to pass the `requestId` in the varargs of `QueryRowEx` instead of using `fmt.Sprintf`.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> Reviewed 1 of 32 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @THardy98)

___
*[-- commits, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtiaiL72EGsRm3WpcoE):*
Is it possible to split this off into its own commit? 
___
*[pkg/server/serverpb/status.proto, line 1282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mti_Px620dPaNMQHQ0T:-Mti_Px620dPaNMQHQ0U:b-t2unp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/server/serverpb/status.proto#L1282)):*
> ```protobuf
> 
> message CancelStatementDiagnosticsReportRequest {
>   int64 request_id = 1;
> ```

Changing protobuf field can cause backward/forward compatibility issues. We should really never change the datatype (as to our best ability, the semantics of those fields) of the protobuf message. 

In this case, we should introduce additional field with field number = 2 and mark the old field as reserved. 

Also I think you can use `gogoproto.customname`  to override the generated code to have the name `RequestID` instead of `RequestId` 
___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtibbThCkZ5yWv4hB0n:-MtibbThCkZ5yWv4hB0o:bd2xi25) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx#L114)):*
> ```XML
>                   return {
>                     name: (
>                       <div className={cx("download-diagnostics-link")}>
> ```

Hmm, not sure if this is common but the CSS class here is used for download link instead of a cancel button, it looks odd to me.

Also, it wasn't quite clear to me, how is the `onClick` implemented for this `div` component ? 
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtibzwRAUzRiEV_k2_4:-MtibzwRAUzRiEV_k2_5:bz0qxnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L58)):*
> ```XML
> 
>     yield put(actions.cancelReportCompleted());
>     yield put(actions.request());
> ```

Hmm why are we putting another request action here? 
___
*[pkg/ui/workspaces/db-console/src/redux/analyticsSagas.ts, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MticG9B-qGno4gwsjN5:-MticG9B-qGno4gwsjN6:bithoc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/db-console/src/redux/analyticsSagas.ts#L36)):*
> ```XML
>   TRACK_CANCEL_DIAGNOSTIC_BUNDLE,
> } from "./analyticsActions";
> import trackCancelDiagnosticsBundle from "oss/src/util/analytics/trackCancelDiagnosticsBundle";
> ```

nit: i think you can just import from `src/.*` without the `oss` prefix. Ditto for the rest of the PR 


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> Reviewed 26 of 32 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @THardy98)

___
*[pkg/server/serverpb/status.proto, line 1282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mti_Px620dPaNMQHQ0T:-Mtjep9q6jsxByXmdfDN:bthpnic) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/server/serverpb/status.proto#L1282)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Changing protobuf field can cause backward/forward compatibility issues. We should really never change the datatype (as to our best ability, the semantics of those fields) of the protobuf message. 

In this case, we should introduce additional field with field number = 2 and mark the old field as reserved. 

Also I think you can use `gogoproto.customname`  to override the generated code to have the name `RequestID` instead of `RequestId` 
</blockquote></details>

In general I would also say to not make this change, but this is a brand new code (added in Dec), and it's not on any release, and is not even being used on the UI, it was created precisely to be used on this issue Thomas is working now, so is safe to make this particular change


<!-- Sent from Reviewable.io -->

Commit:
sql: pass statement request id to backend instead of statement fingerprint

Added a small code change to the cancel statement diagnostics request
API on the backend. Previously, the frontend passed a statement
fingerprint to the API to update the corresponding statement diagnostic
request to an expired state. Now we pass the statement diagnostic ID
directly. This resolves a SQL parsing error that occurs when a statement
fingerprint contains single-quotes (i.e. when a statement fingerprint
contains placeholders).

Release note: None

Commit:
ui: add ability to cancel statement diagnostics to the frontend

Previously, there did not exist an option to cancel a running request
for statement diagnostics, this change provides a cancellation option.

The statements page now has an ellipsis dropdown button where the
download bundle links used to appear. The dropdown contains a
cancellation button (if there is a currently waiting diagnostic request)
and the download links to previously completed diagnostic requests.

The statement details page now has a "Cancel request" button while a
diagnostic request is not completed (i.e. in the "Waiting" status).

The statement diagnostics history page on db-console also has a "Cancel
Request" button while a diagnostic request is not completed (i.e. in the
"Waiting" status).

Added a small code change in the statement diagnostic "Activate" modal.
The expiry time picker allowed for values that were <1, the code change
fixes this bug.

Release note (ui change): Added an option to cancel a running request
for statement diagnostics.

THardy98 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)

___
*[-- commits, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtiaiL72EGsRm3WpcoE):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Is it possible to split this off into its own commit? 
</blockquote></details>

Split the sql/server and the ui changes. Done.
___
*[pkg/server/serverpb/status.proto, line 1282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mti_Px620dPaNMQHQ0T:-MtnA8xQEcU_-4QrC_7J:bp6o2fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/server/serverpb/status.proto#L1282)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

In general I would also say to not make this change, but this is a brand new code (added in Dec), and it's not on any release, and is not even being used on the UI, it was created precisely to be used on this issue Thomas is working now, so is safe to make this particular change
</blockquote></details>

With this in mind I'll keep this change (and keep in mind to be much more conservative with `.proto` changes in the future. Appreciate the feedback, good to learn!). 
___
*[pkg/sql/stmtdiagnostics/statement_diagnostics.go, line 405 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtiduw4Bz6klaIim6BJ:-MtmtBycCj3hmznWPGBP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/sql/stmtdiagnostics/statement_diagnostics.go#L405)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

While we're here, it's safer (from a SQL-injection perspective) to pass the `requestId` in the varargs of `QueryRowEx` instead of using `fmt.Sprintf`.
</blockquote></details>

Done.
___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtibbThCkZ5yWv4hB0n:-Mtmo40c4-G6e7m7Vw6F:bn0mv94) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx#L114)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm, not sure if this is common but the CSS class here is used for download link instead of a cancel button, it looks odd to me.

Also, it wasn't quite clear to me, how is the `onClick` implemented for this `div` component ? 
</blockquote></details>

I believe `download-diagnostics-link` just stylizes colour, text decoration, and changes to text decoration and colour when you hover. Maybe it should be renamed to something more general like `diagnostic-report-dropdown-optio`?

`onClick` for the `div` components is actually handled by the parent `Dropdown` component, using the function `onChange` provided as `props` on line 142. `onChange` is the function called when dropdown items are clicked. For reference (in `dropdown.tsx`):

```
<DropdownItem<T>
        value={menuItem.value}
        onClick={this.handleItemSelection}
        key={idx}
        disabled={menuItem.disabled}
        className={itemsClassname}
      >
        {menuItem.name}
</DropdownItem>
```
and
```
  handleItemSelection = (value: T): void => {
    this.props.onChange(value);
    this.handleMenuOpen();
  };
```

The Dropdown component takes a single function (`onChange`), which is used for all its items `onClick` event, instead of each item having its own `onClick`, it's a little bit unintuitive. I imagine there are many cases where the dropdown items want to do the same thing `onClick`, but scenarios where dropdown items differ with their `onClick` events might prove to be a bit frustrating and result in bloated `onChange` functions to handle each differing event.
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtibzwRAUzRiEV_k2_4:-Mtmmalr6mNxdkiWGi9x:b-wn76bb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L58)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm why are we putting another request action here? 
</blockquote></details>

It's effectively the same as:

```
    yield put(cancelStatementDiagnosticsReportCompleteAction());

    yield put(invalidateStatementDiagnosticsRequests());
    // PUT expects action with `type` field which isn't defined in `refresh` ThunkAction interface
    yield put(refreshStatementDiagnosticsRequests() as any);
```

we want to refresh our statement diagnostics requests after we've cancelled. Update state and UI to reflect how the diagnostic request has changed (i.e. is not longer being requested).

Not related, but this is what I think should be in-place to update state when statement diagnostic requests expire (refresh on a timer set to when the request expires). Currently we have to refresh the page to update the UI on expired requests.
___
*[pkg/ui/workspaces/db-console/src/redux/analyticsSagas.ts, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MticG9B-qGno4gwsjN5:-Mtn9wLP59MFVsdAwQqs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/db-console/src/redux/analyticsSagas.ts#L36)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: i think you can just import from `src/.*` without the `oss` prefix. Ditto for the rest of the PR 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag, @matthewtodd, and @THardy98)

___
*[-- commits, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtiaiL72EGsRm3WpcoE):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

Split the sql/server and the ui changes. Done.
</blockquote></details>

Nice!
___
*[-- commits, line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtn_CGMBG-9FrtB63uc):*
very small nit: let's start the commit msg with "Previsouly, ..." 
___
*[pkg/server/serverpb/status.proto, line 1282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mti_Px620dPaNMQHQ0T:-MtnZdU42b8iUBuYUQpO:bxin1r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/server/serverpb/status.proto#L1282)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

With this in mind I'll keep this change (and keep in mind to be much more conservative with `.proto` changes in the future. Appreciate the feedback, good to learn!). 
</blockquote></details>

Ah thanks for the context! Nice! 
___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/diagnostics/diagnosticsView.tsx, line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtnfZFb964yXGMZNNuA:-MtnfZFb964yXGMZNNuB:b26rh9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/statementDetails/diagnostics/diagnosticsView.tsx#L153)):*
> ```XML
>       width: "160px",
>       render: ((
>         onDownloadDiagnosticBundleClick: (s: string) => void,
> ```

Hmm second thought, is this parameterized callback here necessary? Feels a bit redundant no? We can simply just pull the callbacks into the closure:

I'm thinking something like just using a partial destructuring 


```js
render: () => {
   const { onDownloadDiagnosticsBundleClick, onDiagnosticsCancelRequestClick } = this.props;
}
```
___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtibbThCkZ5yWv4hB0n:-Mtna1-F10S0Uyj_n4oX:bmwtgyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx#L114)):*
>  Maybe it should be renamed to something more general like diagnostic-report-dropdown-optio?

I like this idea a lot better

> The Dropdown component takes a single function (onChange), which is used for all its items onClick event,

I see. That makes a lot more sense. Thanks for explaining


// Also unrelated, it seems like reviewable is so broken after you split off the commit. :facepalm: Uragh i'm resorting back to use github reviews lol.
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtibzwRAUzRiEV_k2_4:-MtndkBh8wKGxLMdg6mc:b6dc3pj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L58)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

It's effectively the same as:

```
    yield put(cancelStatementDiagnosticsReportCompleteAction());

    yield put(invalidateStatementDiagnosticsRequests());
    // PUT expects action with `type` field which isn't defined in `refresh` ThunkAction interface
    yield put(refreshStatementDiagnosticsRequests() as any);
```

we want to refresh our statement diagnostics requests after we've cancelled. Update state and UI to reflect how the diagnostic request has changed (i.e. is not longer being requested).

Not related, but this is what I think should be in-place to update state when statement diagnostic requests expire (refresh on a timer set to when the request expires). Currently we have to refresh the page to update the UI on expired requests.
</blockquote></details>

I see. That makes sense. Interesting point here on changing the state directly rather than go through the request sagas again to get the most up-to-date data. cc @matthewtodd thoughts? 
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtnhz0W28oHJAufUHen:-Mtnhz0W28oHJAufUHeo:b-lx9nvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L60)):*
> ```XML
>     yield put(actions.request());
>   } catch (e) {
>     yield put(actions.cancelReportFailed(e));
> ```


Also on the db console saga you added some additional actions to control the alert message, which is missing in the cluster-ui version of the sagas. Should we add those things here too as well? 
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtne9Yw5eWa10OTgovn:-Mtne9Yw5eWa10OTgovo:bgzj1sa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L95)):*
> ```XML
>     takeEvery(actions.createReport, createDiagnosticsReportSaga),
>     takeLatest(actions.received, receivedStatementsDiagnosticsSaga, delayMs),
>   ]);
> ```

I think we should also update the `all` sagas here to intercept the `cancelReport` actions with the newly added sagas right? This should have been caught by the tests tbh, which is surprising. 
___
*[pkg/ui/workspaces/db-console/src/redux/alerts.ts, line 400 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtnhU2b1Dcz4iXnOcyj:-MtnhU2b1Dcz4iXnOcyk:b-kg5n1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/db-console/src/redux/alerts.ts#L400)):*
> ```XML
>             cancelStatementDiagnosticsAlertLocalSetting.set({ show: false }),
>           );
>           return Promise.resolve();
> ```

Hmmm .. why do we need to resolve the promise here?
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtneXLaBl-x9osZ1eks:-MtneXLaBl-x9osZ1ekt:b-f8p653) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts#L164)):*
> ```XML
>     it("failed request", () => {
>       const error = new Error("Failed request");
>       expectSaga(cancelDiagnosticsReportSaga, actions.cancelReport(request))
> ```

Ok I see why the missing sagas is not being caught by the test, you need to `return` the `expectSaga` for the test to be executed. Otherwise it's just no-op and this is actually just deadcode.


<!-- Sent from Reviewable.io -->

THardy98 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)

___
*[-- commits, line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtn_CGMBG-9FrtB63uc):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

very small nit: let's start the commit msg with "Previsouly, ..." 
</blockquote></details>

Done.
___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/diagnostics/diagnosticsView.tsx, line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtnfZFb964yXGMZNNuA:-Mto1JtGEfAl2BtVEr4T:b9mvi74) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/statementDetails/diagnostics/diagnosticsView.tsx#L153)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm second thought, is this parameterized callback here necessary? Feels a bit redundant no? We can simply just pull the callbacks into the closure:

I'm thinking something like just using a partial destructuring 


```js
render: () => {
   const { onDownloadDiagnosticsBundleClick, onDiagnosticsCancelRequestClick } = this.props;
}
```
</blockquote></details>

Yup, didn't notice that, nice catch. Done.
___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtibbThCkZ5yWv4hB0n:-MtoLa1zASOUAUqwd_Ay:b-64fwui) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5848befd04bb065885f93dd0420a8d5eda4aafb/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTableContent.tsx#L114)):*
> diagnostic-report-dropdown-optio
Renamed the CSS style.
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtne9Yw5eWa10OTgovn:-MtryGog8Vzj5BiZflRm:b20c5tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L95)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

I think we should also update the `all` sagas here to intercept the `cancelReport` actions with the newly added sagas right? This should have been caught by the tests tbh, which is surprising. 
</blockquote></details>

Absolutely, should've caught that myself. 
___
*[pkg/ui/workspaces/db-console/src/redux/alerts.ts, line 400 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtnhU2b1Dcz4iXnOcyj:-Mtryzgr3itV5lOa_wd7:b-eeev3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/db-console/src/redux/alerts.ts#L400)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmmm .. why do we need to resolve the promise here?
</blockquote></details>

I believe the `dismiss` field of the alert needs to return a `Promise` and the return type of the `dispatch` call by itself isn't of type `Promise`.
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtneXLaBl-x9osZ1eks:-Mts3Gu6DQocItS4dFNx:byeanyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts#L164)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Ok I see why the missing sagas is not being caught by the test, you need to `return` the `expectSaga` for the test to be executed. Otherwise it's just no-op and this is actually just deadcode.
</blockquote></details>

From what I've read on the GitHub README for the testing package we use `redux-saga-testing-plan` (https://github.com/jfairbank/redux-saga-test-plan), adding a `return` just lets the test runner know when the `Promise` from the `expectSaga` is resolved (if I understand what I read correctly...). I've played around with the cancel tests, supplying the wrong saga to `expectSaga` and it fails as it should, even without a `return` clause. Again, if I understand correctly from the README, with `expectSaga` we are supplying an action to a saga and testing whether or not the saga does what we anticipate it to do with the action provided. But, I don't think `expectSaga` is designed to test whether our saga is "hooked up" to its corresponding action. Maybe this should be a test on the parent saga `statementsDiagnosticsSagas`, where we supply an action an expect a saga to be executed?


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> Reviewed 2 of 32 files at r1, 1 of 7 files at r2, 1 of 3 files at r4, 3 of 8 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag, @matthewtodd, and @THardy98)

___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtnhz0W28oHJAufUHen:-MtsQAIB300mT6fP4F5T:b-zf3kvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L60)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Also on the db console saga you added some additional actions to control the alert message, which is missing in the cluster-ui version of the sagas. Should we add those things here too as well? 
</blockquote></details>

Follow up here, any thoughts on adding alerts sagas stuff here for the cluster ui? 
___
*[pkg/ui/workspaces/db-console/src/redux/alerts.ts, line 400 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtnhU2b1Dcz4iXnOcyj:-MtsKHAH59ekybcH-8Fk:bntqffx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/db-console/src/redux/alerts.ts#L400)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

I believe the `dismiss` field of the alert needs to return a `Promise` and the return type of the `dispatch` call by itself isn't of type `Promise`.
</blockquote></details>

Ah I see. I thought `Promise.resolve()` returns a concrete value resolved from the `Promise` it's resolving, but reading the docs seems like it's just returning a resolved `Promise` :sweat_smile: 
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtneXLaBl-x9osZ1eks:-MtsPRXs2IMkErvB_jX1:baog234) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts#L164)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

From what I've read on the GitHub README for the testing package we use `redux-saga-testing-plan` (https://github.com/jfairbank/redux-saga-test-plan), adding a `return` just lets the test runner know when the `Promise` from the `expectSaga` is resolved (if I understand what I read correctly...). I've played around with the cancel tests, supplying the wrong saga to `expectSaga` and it fails as it should, even without a `return` clause. Again, if I understand correctly from the README, with `expectSaga` we are supplying an action to a saga and testing whether or not the saga does what we anticipate it to do with the action provided. But, I don't think `expectSaga` is designed to test whether our saga is "hooked up" to its corresponding action. Maybe this should be a test on the parent saga `statementsDiagnosticsSagas`, where we supply an action an expect a saga to be executed?
</blockquote></details>

Hmm interesting that it would still fail without return. But anyhow, looking at the API for `expectSagas.run()`, it returns a `Promise<RunResult>`, which in itself is a promise. Then looking at the `jest` [docs](https://jestjs.io/docs/asynchronous#promises) (our test runner for cluster-ui), it states that: 

"Be sure to return the promise - if you omit this return statement, your test will complete before the promise returned from fetchData resolves and then() has a chance to execute the callback."


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
Changes looking good, can you add some images of before/after here?

> Reviewed 1 of 32 files at r1, 5 of 7 files at r2, 1 of 1 files at r3, 2 of 3 files at r4, 7 of 8 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-:-MtxO2TWEMxbTtTivxnN:b-ql22oo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd and @THardy98)



<!-- Sent from Reviewable.io -->

THardy98 reviewed (commented):
Added a Loom showing the changes to DB Console. Will add another for Serverless/Dedicated when I can get it running with my local UI.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-:-MuBQjkQ4BWUL9P9yvPT:b759j33)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)



<!-- Sent from Reviewable.io -->

THardy98 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)

___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtnhz0W28oHJAufUHen:-MuCbrnI-V6ifv4qIohT:b-xhqbou) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L60)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Follow up here, any thoughts on adding alerts sagas stuff here for the cluster ui? 
</blockquote></details>

I've been trying to get these UI changes to show on Serverless so I could see if maybe some sort of alerting exists already, but I've been having trouble setting it up.

Aside from that, from what I can tell, it seems like cluster-ui does provide `NotificationMessage`s that seem similar to the `Alert`s provided by db-console. I could try adding them, but I don't really have a way of testing/viewing the results right now...
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-MtneXLaBl-x9osZ1eks:-MuCbYbm6U-G2LvAAmTT:b1lxbn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts#L164)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm interesting that it would still fail without return. But anyhow, looking at the API for `expectSagas.run()`, it returns a `Promise<RunResult>`, which in itself is a promise. Then looking at the `jest` [docs](https://jestjs.io/docs/asynchronous#promises) (our test runner for cluster-ui), it states that: 

"Be sure to return the promise - if you omit this return statement, your test will complete before the promise returned from fetchData resolves and then() has a chance to execute the callback."
</blockquote></details>

I've added `return`s to the `expectSaga`s. Considering they seem to be working in either case, it's probably just better to err on the side of caution.


<!-- Sent from Reviewable.io -->

THardy98 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)

___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtnhz0W28oHJAufUHen:-MuHMkpw--op103xKNTr:bkgyvdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L60)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

I've been trying to get these UI changes to show on Serverless so I could see if maybe some sort of alerting exists already, but I've been having trouble setting it up.

Aside from that, from what I can tell, it seems like cluster-ui does provide `NotificationMessage`s that seem similar to the `Alert`s provided by db-console. I could try adding them, but I don't really have a way of testing/viewing the results right now...
</blockquote></details>

From inspecting alerts on CC console, it seems CC console uses `Notification` from `notification.tsx` in managed service to display the same thing an `Alert` does on db-console (the styling classes being used on the element I could only find in managed service used by the `Notification` component).
With db-console, we hook the alert as part of the db-console saga when we cancel/create a statement diagnostic. With CC console, the component belongs to managed service, but the saga belongs to cluster-ui. I don't think we can import a managed service component into cluster-ui (correct me if I'm wrong) so I don't think hooking the component into the cluster-ui saga is feasible.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
Nice! Do you mind also record another video showing this change on the CC Console? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-:-MuMQEvh8iBRur9hbik6:b-qgwdci)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)

___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtnhz0W28oHJAufUHen:-MuMPSiX5wDGJAyowksz:b-vcip7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L60)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

From inspecting alerts on CC console, it seems CC console uses `Notification` from `notification.tsx` in managed service to display the same thing an `Alert` does on db-console (the styling classes being used on the element I could only find in managed service used by the `Notification` component).
With db-console, we hook the alert as part of the db-console saga when we cancel/create a statement diagnostic. With CC console, the component belongs to managed service, but the saga belongs to cluster-ui. I don't think we can import a managed service component into cluster-ui (correct me if I'm wrong) so I don't think hooking the component into the cluster-ui saga is feasible.
</blockquote></details>

Hmm I see. This is tricky. I think it might be a good time for us to think about how to merge `Alert` and `Notification` into a single sharing component that lives within `cluster-ui`, which can be both imported by `db-console` and `managed-service`. Of course, this would be a separate PR. 

Do you mind open an issue on this and leave a TODO here noting that the alerting sagas is missing, and link the issue number here. 


<!-- Sent from Reviewable.io -->

THardy98 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)

___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts, line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-Mtnhz0W28oHJAufUHen:-MuMgDT80GLndpTYtwdD:bf2kp2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bb513ca69bbd84f315cba778f17a1b9c00580a7/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.ts#L60)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm I see. This is tricky. I think it might be a good time for us to think about how to merge `Alert` and `Notification` into a single sharing component that lives within `cluster-ui`, which can be both imported by `db-console` and `managed-service`. Of course, this would be a separate PR. 

Do you mind open an issue on this and leave a TODO here noting that the alerting sagas is missing, and link the issue number here. 
</blockquote></details>

Issue: https://github.com/cockroachdb/cockroach/issues/75559

Added TODO in `statementDiagonostics.sagas.ts`.


<!-- Sent from Reviewable.io -->

THardy98 reviewed (commented):
I've added a quick Loom for CC Console changes with a cluster using fake console & intrusion backends. 

I'm not sure whether this is sufficient to test the changes. With fake backends we still trigger the cluster-ui cancel statement diagnostic saga and test that it's working correctly on CC console, but we don't test it with a real cluster e2e.

The issue I've been having, is creating a cluster on v22.1 that allows for local db backend changes. To create a real cluster on v22.1, you can supply a `version` to the `admin-cli create cluster` command. This uses a docker image with the specified version from our GCP repository of docker images as the cluster's backend. However, local db backend changes won't be represented in this docker image, which is a pain for testing changes that include cloud frontend and db backend.

I have talked with @matthewtodd about this and he suggested separating the backend & frontend changes. We could get the backend merged today, and a new docker image with these changes might be available by Monday, that would allow for testing the frontend changes in a separate PR. Thoughts? @azhng @maryliag 

**UPDATE**: created separate PRs for the [backend](https://github.com/cockroachdb/cockroach/pull/75732) and [frontend](https://github.com/cockroachdb/cockroach/pull/75733) changes. Closing this PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74818#-:-MuWe5er5xWhKI6v37sT:bf6c3t0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, and @THardy98)



<!-- Sent from Reviewable.io -->

maryliag commented:
If this PR is no longer valid, you should close it
